File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/QuestionnaireResource.java
package uk.gov.justice.hmpps.prison.api.resource;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import uk.gov.justice.hmpps.prison.api.model.ErrorResponse;
import uk.gov.justice.hmpps.prison.api.model.Questionnaire;
import uk.gov.justice.hmpps.prison.service.IncidentService;

import javax.validation.constraints.NotNull;

@RestController
@Tag(name = "questionnaires")
@Validated
@RequestMapping("${api.base.path}/questionnaires")
public class QuestionnaireResource {
    private final IncidentService incidentService;

    public QuestionnaireResource(final IncidentService incidentService) {
        this.incidentService = incidentService;
    }

    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "OK", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = Questionnaire.class))}),
            @ApiResponse(responseCode = "400", description = "Invalid request.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))}),
            @ApiResponse(responseCode = "404", description = "Requested resource not found.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))}),
            @ApiResponse(responseCode = "500", description = "Unrecoverable error occurred whilst processing request.", content = {@Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))})})
    @Operation(summary =  "Return a questionnaire for a specified category and code")
    @GetMapping("/{category}/{code}")
    public Questionnaire getQuestionnaire(@PathVariable("category") @Parameter(description = "Category", example = "IR_TYPE", required = true) @NotNull final String category, @PathVariable("code") @Parameter(description = "Code", example = "ASSAULT", required = true) @NotNull final String code) {
        return incidentService.getQuestionnaire(category, code);
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/QuestionnaireResource.kt
package uk.gov.justice.hmpps.prison.api.resource

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpStatus
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import uk.gov.justice.hmpps.prison.api.model.ErrorResponse
import uk.gov.justice.hmpps.prison.api.model.Questionnaire
import uk.gov.justice.hmpps.prison.core.SlowReportQuery
import uk.gov.justice.hmpps.prison.service.QuestionnaireService

@RestController
@Tag(name = "questionnaire")
@Validated
@RequestMapping(value = ["/api/questionnaire"], produces = ["application/json"])
class QuestionnaireResource(
  private val questionnaireService: QuestionnaireService,
) {

  @GetMapping
  @ResponseStatus(HttpStatus.OK)
  @Operation(
    summary = "Returns questionnaire data sets for incident reports",
    description = "No additional role required",
    responses = [
      ApiResponse(
        responseCode = "200",
        description = "Returns questionnaires",
      ),
      ApiResponse(
        responseCode = "401",
        description = "Unauthorized to access this endpoint",
        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
      ),
      ApiResponse(
        responseCode = "403",
        description = "Missing privileges to access this endpoint",
        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
      ),
    ],
  )
  @SlowReportQuery
  fun getQuestionnaires(
    @RequestParam(name = "code", required = false) code: String? = null,
  ): List<Questionnaire> =
    questionnaireService.getQuestionnaires(questionnaireCode = code)
}
